### PR TITLE
Sync components

### PR DESCRIPTION
**Pull request by denise-opendaq**
## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Description:

### CODE REVIEW

File: bindings/python/opendaq/generated/device/py_device.cpp [hunk 1]
```diff
@@ -69,7 +69,7 @@ void defineIDevice(pybind11::module_ m, PyDaqIntf<daq::IDevice, daq::IFolder> cl
             return objectPtr.getCustomComponents().detach();
         },
         py::return_value_policy::take_ownership,
-        "Gets a list of all components/folders in a device that are not titled 'io', 'sig', 'dev' or 'fb'");
+        "Gets a list of all components/folders in a device that are not titled 'io', 'sig', 'dev', 'Synchronization' or 'fb'");
```
Comment (JakaMohorkoDS) [inline bindings/python/opendaq/generated/device/py_device.cpp:72]: Invalid capitalization of other folders, would be great if we could fix this.

File: core/opendaq/device/include/opendaq/device_impl.h [hunk 2]
```diff
@@ -183,10 +186,12 @@ GenericDevice<TInterface, Interfaces...>::GenericDevice(const ContextPtr& ctx,
 {
     this->defaultComponents.insert("Dev");
     this->defaultComponents.insert("IO");
+    this->defaultComponents.insert("Synchronization");
```
Comment (JakaMohorkoDS) [inline core/opendaq/device/include/opendaq/device_impl.h:189]: Could we maybe try if things work properly over OPC UA if we rename this to "Sync"?

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+template <typename MainInterface, typename ... Interfaces>
+GenericSyncComponentImpl<MainInterface, Interfaces...>::GenericSyncComponentImpl(const ContextPtr& context,
+                             const ComponentPtr& parent,
+                             const StringPtr& localId,
+                             const StringPtr& className,
+                             const StringPtr& name)
+    : Super(context, parent, localId, className, name)
+{
+    Super::addProperty(ObjectProperty("Interfaces", PropertyObject()));
+    Super::addProperty(ListProperty("InterfaceNames", List<IString>()));
```
Comment (JakaMohorkoDS) [inline core/opendaq/synchronization/include/opendaq/sync_component_impl.h:78]: This should probably be read-only? It's likely modified by default once an interface is added.

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+template <typename MainInterface, typename ... Interfaces>
+template <typename T>
+typename InterfaceToSmartPtr<T>::SmartPtr GenericSyncComponentImpl<MainInterface, Interfaces...>::getTypedProperty(const StringPtr& name)
+{
+    return this->objPtr.getPropertyValue(name).template asPtr<T>();
+}
+
+template <typename MainInterface, typename ... Interfaces>
+ErrCode GenericSyncComponentImpl<MainInterface, Interfaces...>::getSyncLocked(Bool* synchronizationLocked)
```
Comment (JakaMohorkoDS) [inline core/opendaq/synchronization/include/opendaq/sync_component_impl.h:92]: Implementations are missing OPENDAQ_PARAM_NOT_NULL checks.

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+template <typename MainInterface, typename ... Interfaces>
+ErrCode GenericSyncComponentImpl<MainInterface, Interfaces...>::addInterface(IPropertyObject* syncInterface)
+{
+    OPENDAQ_PARAM_NOT_NULL(syncInterface);
+
+    PropertyObjectPtr interfacePtr = syncInterface;
+
+    //TBD: Check if interface inherits from SyncInterfaceBase
```
Comment (JakaMohorkoDS) [inline core/opendaq/synchronization/include/opendaq/sync_component_impl.h:168]: Yes, the class should inherit from SyncInterfaceBase.

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        auto parentName = objectClass.getParentName();
+        if (!parentName.assigned() || parentName != "SyncInterfaceBase")
+        {
+            return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' does not inherit from 'SyncInterfaceBase'.", className));
+        }
+    }
+    else
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' is not IPropertyObjectClass", className));
+    }
```
Comment (JakaMohorkoDS) [inline core/opendaq/synchronization/include/opendaq/sync_component_impl.h:199]: This should check recursively up the class inheritance chain for whether the base class is "SyncInterfaceBase"

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    err = InterfacesPtr->removeProperty(interfaceName);
+    if (OPENDAQ_FAILED(err))
+        return err;
+
+    return daqTry([&]() {
+        Int selectedSource = 0;
+        getSelectedSource(&selectedSource);
+
+        ListPtr<IString> interfaceNames = getTypedProperty<IList>("InterfaceNames");
+        for (SizeT i = 0; i < interfaceNames.getCount(); i++)
```
Comment (JakaMohorkoDS) [inline core/opendaq/synchronization/include/opendaq/sync_component_impl.h:241]: What happens when the list is empty?

File: core/opendaq/synchronization/include/opendaq/sync_component_internal.h [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * as an input, while others can be used as sync outputs to synchronize other devices.
+ * The configuration of these interfaces and the reading of their status is defined in Part 4.
+ *
+ * @note Depending on the setup, some interfaces may be switched off, and some interfaces may
+ * act as sync sources or outputs.
+ *
+ * @note A CLK interface can be used to let a device run in Fre-Run mode, where the device
+ * syncs internally to an internal quartz.
+ */
+DECLARE_OPENDAQ_INTERFACE(ISyncComponentInternal, IBaseObject)
```
Comment (JakaMohorkoDS) [inline core/opendaq/synchronization/include/opendaq/sync_component_internal.h:54]: This should be ISyncComponentPrivate.

File: modules/ref_device_module/src/ref_device_impl.cpp [hunk 9]
```diff
@@ -127,11 +128,14 @@ void RefDeviceImpl::initIoFolder()
...
-    syncComponent.getOnPropertyValueWrite("UseSync") +=
-        [this](PropertyObjectPtr& obj, PropertyValueEventArgsPtr& args) { };
+    SyncComponentPtr syncComponent;
+    this->getSyncComponent(&syncComponent);
+    SyncComponentInternalPtr syncComponentInternal = syncComponent.asPtr<ISyncComponentInternal>(true);
+
+    syncComponentInternal.addInterface(PropertyObject(this->context.getTypeManager(), "PtpSyncInterface"));
+    syncComponentInternal.addInterface(PropertyObject(this->context.getTypeManager(), "InterfaceClockSync"));
+    syncComponent.setSelectedSource(1);
+    syncComponentInternal.setSyncLocked(true);
```
Comment (JakaMohorkoDS) [inline modules/ref_device_module/src/ref_device_impl.cpp:138]: We should probably keep an example of a custom component.

**Review by JakaMohorkoDS — CHANGES_REQUESTED** [review]
Looks good in general, some smaller comments here and there.

I'd also do 2 more things:
- Validate that it works with the 3.0 simulator image
- Show the sync component in the Device and Full Topology view in the Python app. Can be modelled the same as custom components. Notably, the "Sync" component should be added as a default component there as well.
- Check the Antora documentation for mentions of default components.

File: core/opendaq/modulemanager/src/context_impl.cpp [hunk 10]
```diff
@@ -217,6 +222,77 @@ ErrCode ContextImpl::getDiscoveryServers(IDict** services)
...
+                                                            {"Priority2", 0},
+                                                            {"Profiles", Enumeration("PtpProfileEnumeration", "I558", typeManager)}}),
+                                                            typeManager)));
+
+    PropertyObjectPtr ports = PropertyObject();
+    parameters.addProperty(ObjectProperty("Ports", ports));
+
+    auto ptpSyncInterface = PropertyObjectClassBuilder(typeManager, "PtpSyncInterface")
+                                    .setParentName("SyncInterfaceBase")
+                                    .addProperty(SelectionProperty("Mode", List<IString>("Input", "Output", "Auto", "Off"), 3))
```
Comment (JakaMohorkoDS) [inline core/opendaq/modulemanager/src/context_impl.cpp:288]: Redundant, as it's inherited from parent.

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h (hunk 7 above)
Comment (denise-opendaq) [reply core/opendaq/synchronization/include/opendaq/sync_component_impl.h:241]: nothing will happened as we are initilizing "InterfaceNames" in constructor with empty list. which in native passes while deserilizing. In opcua its populated in tmsSyncComponent constructor as well

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h (hunk 3 above)
Comment (denise-opendaq) [reply core/opendaq/synchronization/include/opendaq/sync_component_impl.h]: we can change the InterfaceNames on add/remove interface

File: changelog/changelog_3.0.0-4.0.0.txt [hunk 11]
```diff
@@ -1,3 +1,21 @@
+06.08.2024
+Description
+  - Intergration the Sync Component
```
Comment (JakaMohorkoDS) [inline changelog/changelog_3.0.0-4.0.0.txt:3]: We can try a bit harder here :D Explaining what it is and what the required tasks to integrate it are.

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h (hunk 3 above)
Comment (JakaMohorkoDS) [reply core/opendaq/synchronization/include/opendaq/sync_component_impl.h]: Yes, but we do so internally. We can call `setPropertyValueProtected`, while outside users should really not modify this. Honestly, I would even make this property invisible.

File: core/opendaq/synchronization/include/opendaq/sync_component_impl.h (hunk 7 above)
Comment (JakaMohorkoDS) [reply core/opendaq/synchronization/include/opendaq/sync_component_impl.h:267]: Is something to consider in the future, but this implies that anyone using selection properties needs to double check that the index is not out of bounds.

**Review by JakaMohorkoDS — COMMENTED** [review]
Lgtm, some minor comments. Could we add 2 more tests to double check if a custom property is added to the SyncComponent that it's transferred over both OPC UA and Native? So, if we for example add a ObjectProperty named "foo" to the SyncComponent object, it should get transferred to the client.

**Review by JakaMohorkoDS — APPROVED** [review]
Looks good!

**Pull request merged**